src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs

```diff
@@ -103,36 +133,42 @@ internal static void Traverse<T>(this SelectExpandClause selectExpandClause, Fun
             List<string> selectList = selectExpandClause.GetCurrentLevelSelectList();
             List<T> expandList = new List<T>();
 
-            foreach (ExpandedNavigationSelectItem expandSelectItem in selectExpandClause.SelectedItems.Where(I => I.GetType() == typeof(ExpandedNavigationSelectItem)))
+            foreach (SelectItem selectItem in selectExpandClause.SelectedItems)
             {
-                string currentExpandClause = String.Join("/", expandSelectItem.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray());
-                T subResult = default(T);
-                if (expandSelectItem.SelectAndExpand.SelectedItems.Any())
-                {
-                    Traverse(expandSelectItem.SelectAndExpand, processSubResult, combineSelectAndExpand, processApply, out subResult);
-                }
-
-                if (expandSelectItem.ApplyOption != null && processApply != null)
-                {
-                    subResult = processApply(expandSelectItem.ApplyOption);
-                }
+                // $expand=..../$ref
+                ExpandedReferenceSelectItem expandRefItem = selectItem as ExpandedReferenceSelectItem;
+                ExpandedNavigationSelectItem expandSelectItem = selectItem as ExpandedNavigationSelectItem;

```

Comment:
Move this to after the if check. Cast only when expandSelectItem fails. cast inside else i.e.

Reply:
#Resolved